File: java/benchmark/src/main/java/de/hhu/bsinfo/observatory/benchmark/OperationPhase.java
package de.hhu.bsinfo.observatory.benchmark;

import de.hhu.bsinfo.jdetector.lib.IbPerfCounter;
import de.hhu.bsinfo.jdetector.lib.exception.IbFileException;
import de.hhu.bsinfo.jdetector.lib.exception.IbMadException;
import de.hhu.bsinfo.observatory.benchmark.Benchmark.Mode;
import de.hhu.bsinfo.observatory.benchmark.result.LatencyMeasurement;
import de.hhu.bsinfo.observatory.benchmark.result.OverheadMeasurement;
import de.hhu.bsinfo.observatory.benchmark.result.Status;
import de.hhu.bsinfo.observatory.benchmark.result.ThroughputMeasurement;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.TreeMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

class OperationPhase extends BenchmarkPhase {

    private static final Logger LOGGER = LoggerFactory.getLogger(OperationPhase.class);

    private final Operation operation;

    OperationPhase(Benchmark benchmark, Operation operation) {
        super(benchmark);

        this.operation = operation;
    }

    private Status calculateOverhead() {
        IbPerfCounter perfCounter = getBenchmark().getPerfCounter();

        try {
            perfCounter.refreshCounters();
        } catch (IbFileException | IbMadException e) {
            LOGGER.error("Unable to refresh performance counters", e);
            return Status.NETWORK_ERROR;
        }

        if(operation instanceof BidirectionalThroughputOperation || operation instanceof MessagingPingPongOperation) {
            operation.setOverheadMeasurement(new OverheadMeasurement(perfCounter.getXmitDataBytes() +
                    perfCounter.getRcvDataBytes(), operation.getMeasurement()));
        } else if(operation instanceof RdmaReadThroughputOperation || operation instanceof RdmaReadLatencyOperation) {
            operation.setOverheadMeasurement(new OverheadMeasurement(operation.getMode() == Mode.SEND ?
                    perfCounter.getRcvDataBytes() : perfCounter.getXmitDataBytes(), operation.getMeasurement()));
        } else {
            operation.setOverheadMeasurement(new OverheadMeasurement(operation.getMode() == Mode.SEND ?
                    perfCounter.getXmitDataBytes() : perfCounter.getRcvDataBytes(), operation.getMeasurement()));
        }

        return Status.OK;
    }

    private void saveSingleResult(String path, String operationSize, Map<String, Double> valueMap) throws IOException {
        File file = new File(path);

        if(!file.getParentFile().exists()) {
            if(!file.getParentFile().mkdirs()) {
                throw new IOException("Unable to create folder '" + file.getParentFile().getPath() + "'");
            }
        }

        BufferedWriter writer = new BufferedWriter(new FileWriter(file, true));

        if(file.length() == 0) {
            writer.append("Benchmark,Iteration,Size");

            for(String key : valueMap.keySet()) {
                writer.append(",").append(key);
            }

            writer.newLine();
        }

        writer.append(getBenchmark().getResultName()).append(",")
                .append(String.valueOf(getBenchmark().getIterationNumber())).append(",")
                .append(operationSize);

        for(double value : valueMap.values()) {
            writer.append(",").append(String.format("%.12e", value));
        }

        writer.newLine();

        writer.flush();
        writer.close();
    }

    private void saveResults() throws IOException {
        Map<String, Double> valueMap = new TreeMap<>();

        if(operation instanceof ThroughputOperation) {
            ThroughputMeasurement measurement = ((ThroughputOperation) operation).getMeasurement();

            valueMap.put("OperationThroughput", measurement.getOperationThroughput());
            valueMap.put("DataThroughput", measurement.getDataThroughput());
        } else if(operation instanceof LatencyOperation) {
            LatencyMeasurement measurement = ((LatencyOperation) operation).getMeasurement();

            valueMap.put("OperationThroughput", measurement.getOperationThroughput());
            valueMap.put("AverageLatency", measurement.getAverageLatency());
            valueMap.put("MinimumLatency", measurement.getMinimumLatency());
            valueMap.put("MaximumLatency", measurement.getMaximumLatency());
            valueMap.put("50thLatency", measurement.getPercentileLatency(0.5f));
            valueMap.put("95thLatency", measurement.getPercentileLatency(0.95f));
            valueMap.put("99thLatency", measurement.getPercentileLatency(0.99f));
            valueMap.put("999thLatency", measurement.getPercentileLatency(0.999f));
            valueMap.put("9999thLatency", measurement.getPercentileLatency(0.9999f));
        }

        if(getBenchmark().measureOverhead()) {
            valueMap.put("DataOverheadFactor", operation.getOverheadMeasurement().getOverheadFactor());
            valueMap.put("DataOverheadPercentage", operation.getOverheadMeasurement().getOverheadPercentage());
            valueMap.put("DataOverheadThroughput", operation.getOverheadMeasurement().getOverheadDataThroughput());
        }

        saveSingleResult(getBenchmark().getResultPath() + "/" + operation.getOutputFilename() + ".csv",
                String.valueOf(operation.getMeasurement().getOperationSize()), valueMap);
    }

    @Override
    Status execute() {
        LOGGER.info("Executing phase of type '{}' with {} operations of size {} bytes", operation.getClass().getSimpleName(),
                operation.getMeasurement().getOperationCount(), operation.getMeasurement().getOperationSize());

        if(getBenchmark().measureOverhead()) {
            try {
                getBenchmark().getPerfCounter().resetCounters();
            } catch (IbFileException | IbMadException e) {
                LOGGER.error("Unable to reset performance counters", e);
                return Status.NETWORK_ERROR;
            }
        }

        Status status = operation.execute();

        if(status == Status.OK && getBenchmark().measureOverhead()) {
            Status overheadStatus = calculateOverhead();
            if(overheadStatus != Status.OK) {
                LOGGER.error("Measuring overhead failed with status [{}]", overheadStatus);
                return overheadStatus;
            }
        }

        if(status == Status.OK && getBenchmark().isServer()) {
            if(getBenchmark().measureOverhead()) {
                LOGGER.info("Operation finished with results:\n{},\n{}", operation.getMeasurement(), operation.getOverheadMeasurement());
            } else {
                LOGGER.info("Operation finished with results:\n{}", operation.getMeasurement());
            }

            try {
                saveResults();
            } catch (IOException e) {
                LOGGER.error("Unable to save results", e);
                return Status.FILE_ERROR;
            }
        }

        return status;
    }
}

File: cpp/src/observatory/phase/OperationPhase.cpp
#include "OperationPhase.h"
#include <observatory/Benchmark.h>
#include <detector/exception/IbPerfException.h>
#include <observatory/operation/ThroughputOperation.h>
#include <observatory/util/Util.h>
#include <observatory/operation/BidirectionalThroughputOperation.h>
#include <observatory/operation/MessagingPingPongOperation.h>
#include <observatory/operation/RdmaReadThroughputOperation.h>
#include <observatory/operation/RdmaReadLatencyOperation.h>
#include <fstream>
#include <iomanip>

namespace Observatory {

OperationPhase::OperationPhase(Benchmark &benchmark, Operation &operation) :
        BenchmarkPhase(benchmark),
        operation(operation) {}

const char* OperationPhase::getName() {
    return "OperationPhase";
}

Status OperationPhase::execute() {
    LOGGER.info("Executing phase of type '%s' with %u operations of size %u bytes", operation.getClassName(),
                operation.getMeasurement().getOperationCount(), operation.getMeasurement().getOperationSize());

    if(getBenchmark().measureOverhead()) {
        try {
            getBenchmark().getPerfCounter().ResetCounters();
        } catch(Detector::IbPerfException &e) {
            LOGGER.error("Unable to reset performance counters\n\033[0m %s", e.what());
            return Status::NETWORK_ERROR;
        }
    }

    Status status = operation.execute();

    if(status == Status::OK && getBenchmark().measureOverhead()) {
        Status overheadStatus = calculateOverhead();
        if(overheadStatus != Status::OK) {
            LOGGER.error("Measuring overhead failed with status [%s]", getStatusString(overheadStatus));
            return overheadStatus;
        }
    }

    if(status == Status::OK && getBenchmark().isServer()) {
        if(getBenchmark().measureOverhead()) {
            LOGGER.info("Operation finished with results:\n%s,\n%s", static_cast<std::string>(operation.getMeasurement()).c_str(), static_cast<std::string>(operation.getOverheadMeasurement()).c_str());
        } else {
            LOGGER.info("Operation finished with results:\n%s", static_cast<std::string>(operation.getMeasurement()).c_str());
        }

        try {
            saveResults();
        } catch (std::runtime_error &e) {
            LOGGER.error("Unable to save results\n\033[0m %s", e.what());
            return Status::FILE_ERROR;
        }
    }

    return status;
}

Status OperationPhase::calculateOverhead() {
    Detector::IbPerfCounter &perfCounter = getBenchmark().getPerfCounter();

    try {
        perfCounter.RefreshCounters();
    } catch (Detector::IbPerfException &e) {
        LOGGER.error("Unable to refresh performance counters\n\033[0m %s", e.what());
        return Status::NETWORK_ERROR;
    }

    if(Util::instanceof<BidirectionalThroughputOperation>(&operation) || Util::instanceof<MessagingPingPongOperation>(&operation)) {
        operation.setOverheadMeasurement(std::make_shared<OverheadMeasurement>(
                perfCounter.GetXmitDataBytes() + perfCounter.GetRcvDataBytes(), operation.getMeasurement()));
    } else if(Util::instanceof<RdmaReadThroughputOperation>(&operation) || Util::instanceof<RdmaReadLatencyOperation>(&operation)) {
        operation.setOverheadMeasurement(std::make_shared<OverheadMeasurement>(
                operation.getMode() == Benchmark::Mode::SEND ? perfCounter.GetRcvDataBytes() : perfCounter.GetXmitDataBytes(),operation.getMeasurement()));
    } else {
        operation.setOverheadMeasurement(std::make_shared<OverheadMeasurement>(
                operation.getMode() == Benchmark::Mode::SEND ? perfCounter.GetXmitDataBytes() : perfCounter.GetRcvDataBytes(), operation.getMeasurement()));
    }

    return Status::OK;
}

void OperationPhase::saveSingleResult(std::string &path, uint32_t operationSize, std::map<std::string, double> &valueMap) {
    std::ofstream file;
    std::string folderPath = path.substr(0, path.find_last_of('/')) + "'";

    // UGLY!!
    if(std::system(("mkdir -p '" + folderPath).c_str())) {
        throw std::runtime_error("Unable to open file '" + folderPath + "'!");
    }

    file.open(path, std::ios::out | std::ios::app);

    if(!file.is_open()) {
        throw std::runtime_error("Unable to open file '" + path + "'!");
    }

    file.seekp(0, std::ios::end);
    if(file.tellp() == 0) {
        file << "Benchmark,Iteration,Size";

        for(const auto &element : valueMap) {
            file << "," << element.first;
        }

        file << std::endl;
    }

    file << getBenchmark().getResultName() << "," << getBenchmark().getIterationNumber() << "," << operationSize;
    file << std::scientific << std::setprecision(12);

    for(const auto &element : valueMap) {
        file << "," << element.second;
    }

    file << std::endl;
    file.close();
}

void OperationPhase::saveResults() {
    std::string path = getBenchmark().getResultPath() + "/" + operation.getOutputFilename() + ".csv";
    std::map<std::string, double> valueMap;

    if(Util::instanceof<ThroughputOperation>(&operation)) {
        auto &measurement = dynamic_cast<ThroughputMeasurement&>(operation.getMeasurement());
        
        valueMap["OperationThroughput"] = measurement.getOperationThroughput();
        valueMap["DataThroughput"] = measurement.getDataThroughput();
    } else if(Util::instanceof<LatencyOperation>(&operation)) {
        auto &measurement = dynamic_cast<LatencyMeasurement&>(operation.getMeasurement());
        
        valueMap["OperationThroughput"] = measurement.getOperationThroughput();
        valueMap["AverageLatency"] = measurement.getAverageLatency();
        valueMap["MinimumLatency"] = measurement.getMinimumLatency();
        valueMap["MaximumLatency"] = measurement.getMaximumLatency();
        valueMap["50thLatency"] = measurement.getPercentileLatency(0.5f);
        valueMap["95thLatency"] = measurement.getPercentileLatency(0.95f);
        valueMap["99thLatency"] = measurement.getPercentileLatency(0.99f);
        valueMap["999thLatency"] = measurement.getPercentileLatency(0.999f);
        valueMap["9999thLatency"] = measurement.getPercentileLatency(0.9999f);
    }

    if(getBenchmark().measureOverhead()) {
        valueMap["DataOverheadFactor"] = operation.getOverheadMeasurement().getOverheadFactor();
        valueMap["DataOverheadPercentage"] = operation.getOverheadMeasurement().getOverheadPercentage();
        valueMap["DataOverheadThroughput"] = operation.getOverheadMeasurement().getOverheadDataThroughput();
    }

    saveSingleResult(path, operation.getMeasurement().getOperationSize(), valueMap);
}

}